src/util/amount.rs

```diff
@@ -68,22 +76,32 @@ impl FromStr for Denomination {
 
     /// Convert from a str to Denomination.
     ///
-    /// Any combination of upper and/or lower case, excluding uppercase 'M' is considered valid.
-    /// - Singular: BTC, mBTC, uBTC
+    /// Any combination of upper and/or lower case, excluding uppercase of SI(m, u, n, p) is considered valid.
+    /// - Singular: BTC, mBTC, uBTC, nBTC, pBTC
     /// - Plural or singular: sat, satoshi, bit, msat
     ///
-    /// Due to ambiguity between mega and milli we prohibit usage of leading capital 'M'.
+    /// Due to ambiguity between mega and milli, pico and peta we prohibit usage of leading capital 'M', 'P'.
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         use self::ParseAmountError::*;
+        use self::Denomination as D;
+
+        denomination_from_str(s).map_or_else(
```

Comment:
Could this be changed to `match`? I find `map_or_else` less readable.

Reply:
How about this?

```
        match denomination_from_str(s) {
            None => Err(UnknownDenomination(s.to_owned())),
            Some(d @ D::MilliBitcoin) | Some(d @ D::PicoBitcoin) | Some(d @ D::MilliSatoshi) => {
                if s.starts_with(|ch : char| ch.is_uppercase() ) {
                    Err(PossiblyConfusingDenomination(s.to_owned()))
                } else {
                    Ok(d)
                }
            }
            Some(d @ D::NanoBitcoin) | Some(d @ D::MicroBitcoin) => {
                if s.starts_with(|ch : char| ch.is_uppercase() ) {
                    Err(UnknownDenomination(s.to_owned()))
                } else {
                    Ok(d)
                }
            }
            Some(d) => Ok(d),
        }
```

I don't like repeated `if s.starts_with(` but not sure if there is other solution other than nested match?

Reply:
Flattened matches look better to me. Do you have some specific reason for your suggestion?
Using a variable should be fine to get rid of repeating.

Reply:
ahh yours look very nice. I didn't know we can pattern match multiple enum variant with if guard(rust newbie).
I updated code using your suggestion except doing the `starts_with` check inline in if guard so we can save some unnecessary checking in other other cases(`None`,  `Some(d)`). 

Or do u think this minor optimization is not worth the code duplication?

Reply:
FWIW I prefer @Kixunil's version using `let starts_with_uppercase = s.starts_with(|ch : char| ch.is_uppercase());` :)

Reply:
> so we can save some unnecessary checking in other other cases

I was confident that the optimizer would figure it out but checked at godbolt.org and found it wouldn't.
Here's the fix without making the code ugly:

```rust
let starts_with_uppercase = || s.starts_with(|ch : char| ch.is_uppercase());
match denomination_from_str(s) {
    Some(D::MilliBitcoin) | Some(D::PicoBitcoin) | Some(D::MilliSatoshi) if starts_with_uppercase() => {
        Err(PossiblyConfusingDenomination(s.to_owned()))
    },
    Some(D::NanoBitcoin) | Some(D::MicroBitcoin) if starts_with_uppercase() => {
        Err(UnknownDenomination(s.to_owned()))
    },
    Some(denomination) => {
        Ok(denomination)
    },
    None => {
        Err(UnknownDenomination(s.to_owned()))
    },
}
```

It's a bit uncommon style though. :)
Note that anyway calling that function are just two compares and jumps. I suspect that all the time wasted on them for all people running it is less than the amount of time we spent on this but hey, this is fun! :)

Reply:
You are right that it's very trivial optimization. But since this one has the best of both world(it looks like the assembly code generated shorter than my previous clumsy one 😃), we will go for this option then?